src/components/advanced-color-picker/advanced-color-picker.component.tsx

```diff
@@ -11,66 +10,122 @@ import { SimpleColorPicker, SimpleColor } from "../simple-color-picker";
 import { filterStyledSystemMarginProps } from "../../style/utils";
 
-const marginPropTypes = filterStyledSystemMarginProps(
-  styledSystemPropTypes.space
-);
+export interface AdvancedColor {
+  label: string;
+  value: string;
+}
+
+export interface AdvancedColorPickerPropTypes extends MarginProps {
+  /** Prop to specify the aria-describedby property of the component */
+  "aria-describedby"?: string;
+  /**
+   * Prop to specify the aria-label of the component.
+   * To be used only when the title prop is not defined, and the component is not labelled by any internal element.
+   */
+  "aria-label"?: string;
+  /**
+   * Prop to specify the aria-labeledby property of the component
+   * To be used when the title prop is a custom React Node,
+   * or the component is labelled by an internal element other than the title.
+   */
+  "aria-labelledby"?: string;
+  /** Prop for `availableColors` containing array of objects of colors */
+  availableColors: AdvancedColor[];
+  /** Prop for `defaultColor` containing the default color for `uncontrolled` use */
+  defaultColor: string;
+  /** Specifies the name prop to be applied to each color in the group */
+  name: string;
+  /** Prop for `onBlur` event */
+  onBlur?: (ev: React.FocusEvent<HTMLInputElement>) => void;
+  /** Prop for `onChange` event */
+  onChange?: (ev: React.ChangeEvent<HTMLInputElement>) => void;
+  /** Prop for `onClose` event */
+  onClose?: (ev: React.MouseEvent<HTMLElement>) => void;
+  /** Prop for `onOpen` event */
+  onOpen?: (ev: React.MouseEvent<HTMLElement>) => void;
+  /** Prop for `open` status */
+  open?: boolean;
+  /** The ARIA role to be applied to the component container */
+  role?: string;
+  /** Prop for `selectedColor` containing pre-selected color for `controlled` use */
+  selectedColor?: string;
+}
 
 const AdvancedColorPicker = ({
   "aria-describedby": ariaDescribedBy,
   "aria-label": ariaLabel,
   "aria-labelledby": ariaLabelledBy,
+  availableColors,
+  defaultColor,
   name,
-  open,
   onOpen,
   onClose,
   onChange,
   onBlur,
-  availableColors,
-  defaultColor,
-  selectedColor,
+  open,
   role,
+  selectedColor,
   ...props
-}) => {
+}: AdvancedColorPickerPropTypes) => {
   const isOpen = open || false;
-  const [dialogOpen, setDialogOpen] = useState();
+  const [colors, setColors] = useState<
+    { value: string; label: string; getRef: () => HTMLInputElement | null }[]
+  >();
+  const [dialogOpen, setDialogOpen] = useState<boolean>();
   const currentColor = selectedColor || defaultColor;
-  const [selectedColorRef, setSelectedColorRef] = useState();
-
-  const gridItemRefs = useRef(
-    Array.from(
-      {
-        length: availableColors.length,
-      },
-      () => React.createRef()
-    )
+  const [
+    selectedColorRef,
+    setSelectedColorRef,
+  ] = useState<HTMLInputElement | null>(null);
+
+  const simpleColorPickerData = useRef<{
+    gridItemRefs: Array<HTMLInputElement | null>;
+  }>();
+
+  useEffect(
```

Comment:
this is necessary to set proper refs, as the `simpleColorPickerData.current` will be null on initial render, and we're using it's ref list to set the focus on the first element.

When `availableColors` prop change, we will need a re-render anyway, as the main content of the component is based on it, so I can't see a problem here.

Reply:
we're not missing a test as this should be covered by cypress not unit tests

Reply:
as I pointed out, the Cypress tests passed as well 😁 But my point isn't about tests really, I'm saying we should make this change to avoid an unnecessary useEffect - and I'm saying that I have high confidence it is indeed unnecessary because when I rewrite this without the useEffect, all the unit tests and cypress tests for this component still pass.

Reply:
So looks like we do have a missing Cypress test here. Usually I'm leaving that to the QA, as IMO they should do visual tests.

We do need useEffect here unless you have a better way to ensure that the code would be run after the ref is not null.

Reply:
I take that back, you're right. Looks like I resolved the problem when I changed from `ref: simpleColorPickerData.current.gridItemRefs[index]` assignment to `getRef: () => simpleColorPickerData.current.gridItemRefs[index]` method. Too much meddling with these refs...

Reply:
Ah yes, with a `getRef` function that doesn't get called until it's needed it's much less likely (probably impossible) to have a problem with accessing an unitiliased ref.